### PR TITLE
restore-util: support batch split

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

### What problem does this PR solve? 
Support batch split

### What is changed and how it works?
None

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

TODO
- [ ] performance test

### CODE REVIEW

File: pkg/restore-util/client.go [hunk 1]
```diff
@@ -159,6 +162,68 @@ func (c *pdClient) SplitRegion(ctx context.Context, regionInfo *RegionInfo, key
...
+		SplitKeys: keys,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp.RegionError != nil {
+		return nil, errors.Errorf("split region failed: region=%v, err=%v", regionInfo.Region, resp.RegionError)
+	}
+
+	// Assume the new region is the left one.
```
Comment (kennytm) [inline pkg/restore-util/client.go]: Not sure if "left" makes sense in this context 🙃

File: pkg/restore-util/range.go [hunk 2]
```diff
@@ -33,6 +36,47 @@ func (r *Range) contains(key []byte) bool {
...
 
+// sortRanges checks if the range overlapped and sort them
+func sortRanges(ranges []Range, rewriteRules *RewriteRules) ([]Range, error) {
+	rangeTree := NewRangeTree()
+	for _, rg := range ranges {
+		if rewriteRules != nil {
+			var startRule *import_sstpb.RewriteRule
+			rg.StartKey, startRule = replacePrefix(rg.StartKey, rewriteRules)
+			if startRule == nil {
+				log.Warn("cannot find rewrite rule", zap.ByteString("key", rg.StartKey))
```
Comment (kennytm) [inline pkg/restore-util/range.go]:
```suggestion
				log.Warn("cannot find rewrite rule", zap.Binary("key", rg.StartKey))
```

Don't use `ByteString` for bytes which is not UTF-8.

File: pkg/restore-util/range.go [hunk 3]
```diff
@@ -33,6 +36,47 @@ func (r *Range) contains(key []byte) bool {
...
+			} else {
+				log.Debug(
+					"rewrite start key",
+					zap.ByteString("key", rg.StartKey),
```
Comment (kennytm) [inline pkg/restore-util/range.go]:
```suggestion
					zap.Binary("key", rg.StartKey),
```

File: pkg/restore-util/range.go [hunk 4]
```diff
@@ -33,6 +36,47 @@ func (r *Range) contains(key []byte) bool {
...
+					zap.Stringer("rule", startRule))
+			}
+			var endRule *import_sstpb.RewriteRule
+			rg.EndKey, endRule = replacePrefix(rg.EndKey, rewriteRules)
+			if endRule == nil {
+				log.Warn("cannot find rewrite rule", zap.ByteString("key", rg.EndKey))
```
Comment (kennytm) [inline pkg/restore-util/range.go]:
```suggestion
				log.Warn("cannot find rewrite rule", zap.Binary("key", rg.EndKey))
```

File: pkg/restore-util/range.go [hunk 5]
```diff
@@ -33,6 +36,47 @@ func (r *Range) contains(key []byte) bool {
...
+			} else {
+				log.Debug(
+					"rewrite end key",
+					zap.ByteString("key", rg.EndKey),
```
Comment (kennytm) [inline pkg/restore-util/range.go]:
```suggestion
					zap.Binary("key", rg.EndKey),
```

File: pkg/restore-util/range.go [hunk 6]
```diff
@@ -33,6 +36,47 @@ func (r *Range) contains(key []byte) bool {
...
+					zap.Stringer("rule", endRule))
+			}
+		}
+		if out := rangeTree.InsertRange(rg); out != nil {
+			return nil, errors.Errorf("ranges overlapped: %v, %v", out.(*Range).String(), rg.String())
```
Comment (kennytm) [inline pkg/restore-util/range.go]:
```suggestion
			return nil, errors.Errorf("ranges overlapped: %s, %s", out, rg)
```

`%s` would implicitly call `.String()` for any `Stringer`.

File: pkg/restore-util/split_test.go [hunk 7]
```diff
@@ -106,32 +139,33 @@ func (c *testClient) GetOperator(ctx context.Context, regionID uint64) (*pdpb.Ge
...
+	}
+	return regions, nil
 }
 
 // region: [, aay), [aay, bba), [bba, bbh), [bbh, cca), [cca, )
 // range: [aaa, aae), [aae, aaz), [ccd, ccf), [ccf, ccj)
 // rewrite rules: aa -> xx,  cc -> bb
 // expected regions after split:
-// 		[, aay), [aay, bb), [bb, bba), [bba, bbf), [bbf, bbh), [bbh, cca), [cca, xx), [xx, xxe), [xxe, xxz), [xxz, )
+// 	[, aay), [aay, bb), [bb, bba), [bba, bbf), [bbf, bbj), [bbj, bbh] [bbh, cca), [cca, xx), [xx, xxe), [xxe, xxz), [xxz, )
```
Comment (kennytm) [inline pkg/restore-util/split_test.go]:
```suggestion
// 	[, aay), [aay, bb), [bb, bba), [bba, bbf), [bbf, bbh), [bbh, bbj), [bbj, cca), [cca, xx), [xx, xxe), [xxe, xxz), [xxz, )
```

File: pkg/restore-util/split_test.go [hunk 8]
```diff
@@ -217,14 +251,14 @@ func initRewriteRules() *RewriteRules {
 }
 
 // expected regions after split:
-// 		[, aay), [aay, bb), [bb, bba), [bba, bbf), [bbf, bbh), [bbh, cca), [cca, xx), [xx, xxe), [xxe, xxz), [xxz, )
+// 	[, aay), [aay, bb), [bb, bba), [bba, bbf), [bbf, bbj), [bbj, bbh] [bbh, cca), [cca, xx), [xx, xxe), [xxe, xxz), [xxz, )
```
Comment (kennytm) [inline pkg/restore-util/split_test.go]:
```suggestion
// 	[, aay), [aay, bb), [bb, bba), [bba, bbf), [bbf, bbh), [bbh, bbj), [bbj, cca), [cca, xx), [xx, xxe), [xxe, xxz), [xxz, )
```

File: pkg/restore-util/split.go [hunk 9]
```diff
@@ -219,58 +181,68 @@ func (rs *RegionSplitter) waitForScatterRegion(ctx context.Context, regionInfo *
...
+	for _, rg := range ranges {
+		checkKeys = append(checkKeys, rg.EndKey)
+	}
+	for _, key := range checkKeys {
+		if region := needSplit(key, regions); region != nil {
+			_, ok := splitKeyMap[region.Region.GetId()]
+			if !ok {
+				splitKeyMap[region.Region.GetId()] = make([][]byte, 0, 1)
+			}
+			splitKeyMap[region.Region.GetId()] = append(splitKeyMap[region.Region.GetId()], key)
```
Comment (kennytm) [inline pkg/restore-util/split.go]:
```go
splitKeys, ok := splitKeyMap[region.Region.GetId()]
if !ok {
	splitKeys = make([][]byte, 0, 1)
}
splitKeyMap[region.Region.GetId()] = append(splitKeys, key)
```

File: pkg/restore-util/split.go [hunk 10]
```diff
@@ -40,50 +40,71 @@ func NewRegionSplitter(client Client) *RegionSplitter {
...
-	if err != nil {
-		return errors.Trace(err)
+	minKey := codec.EncodeBytes([]byte{}, sortedRanges[0].StartKey)
+	maxKey := codec.EncodeBytes([]byte{}, sortedRanges[len(sortedRanges)-1].EndKey)
+	for _, rule := range rewriteRules.Table {
+		if bytes.Compare(minKey, rule.GetNewKeyPrefix()) > 0 {
+			minKey = rule.GetNewKeyPrefix()
+		}
+		if bytes.Compare(maxKey, rule.GetNewKeyPrefix()) < 0 {
+			maxKey = rule.GetOldKeyPrefix()
```
Comment (kennytm) [inline pkg/restore-util/split.go]:
```suggestion
			maxKey = rule.GetNewKeyPrefix()
```

File: pkg/restore-util/split.go [hunk 11]
```diff
@@ -40,50 +40,71 @@ func NewRegionSplitter(client Client) *RegionSplitter {
...
-		if rg == nil {
-			return false
+	for _, rule := range rewriteRules.Data {
+		if bytes.Compare(minKey, rule.GetNewKeyPrefix()) > 0 {
+			minKey = rule.GetNewKeyPrefix()
 		}
-		if onSplit != nil {
-			onSplit(rg)
+		if bytes.Compare(maxKey, rule.GetNewKeyPrefix()) < 0 {
+			maxKey = rule.GetOldKeyPrefix()
```
Comment (kennytm) [inline pkg/restore-util/split.go]:
```suggestion
			maxKey = rule.GetNewKeyPrefix()
```

File: pkg/restore-util/split.go [hunk 12]
```diff
@@ -40,50 +40,71 @@ func NewRegionSplitter(client Client) *RegionSplitter {
...
+		for regionID, keys := range splitKeyMap {
+			var newRegions []*RegionInfo
+			newRegions, err = rs.splitAndScatterRegions(ctx, regionMap[regionID], keys)
+			if err != nil {
+				interval = 2 * interval
+				if interval > SplitMaxRetryInterval {
+					interval = SplitMaxRetryInterval
+				}
+				time.Sleep(interval)
+				continue SplitRegions
```
Comment (kennytm) [inline pkg/restore-util/split.go]: should the `err` be logged somewhere?

File: pkg/restore-util/split.go [hunk 13]
```diff
@@ -40,50 +40,74 @@ func NewRegionSplitter(client Client) *RegionSplitter {
...
+			var newRegions []*RegionInfo
+			newRegions, err = rs.splitAndScatterRegions(ctx, regionMap[regionID], keys)
+			if err != nil {
+				interval = 2 * interval
+				if interval > SplitMaxRetryInterval {
+					interval = SplitMaxRetryInterval
+				}
+				time.Sleep(interval)
+				if i > 3 {
+					log.Warn("splitting regions failed, retry it")
```
Comment (kennytm) [inline pkg/restore-util/split.go]:
```suggestion
					log.Warn("splitting regions failed, retry it", zap.Error(err))
```

(maybe include more information)

**Review by kennytm — COMMENTED** [review]
LGTM

File: pkg/restore-util/client.go [hunk 14]
```diff
@@ -176,6 +179,68 @@ func (c *pdClient) SplitRegion(ctx context.Context, regionInfo *RegionInfo, key
...
+		peer = regionInfo.Region.Peers[0]
+	}
+
+	storeID := peer.GetStoreId()
+	store, err := c.GetStore(ctx, storeID)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := grpc.Dial(store.GetAddress(), grpc.WithInsecure())
+	defer conn.Close()
```
Comment (overvenus) [inline pkg/restore-util/client.go]: Please check error first.

File: pkg/restore-util/client.go [hunk 15]
```diff
@@ -176,6 +179,68 @@ func (c *pdClient) SplitRegion(ctx context.Context, regionInfo *RegionInfo, key
...
+			RegionId:    regionInfo.Region.Id,
+			RegionEpoch: regionInfo.Region.RegionEpoch,
+			Peer:        peer,
+		},
+		SplitKeys: keys,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp.RegionError != nil {
```
Comment (overvenus) [inline pkg/restore-util/client.go:215]: What if `resp` is nil?

Comment (5kbpers) [reply pkg/restore-util/client.go:215]: If `err` is not nil, `resp` should never be nil

**Review by overvenus — APPROVED** [review]
LGTM

**Pull request merged**